Add get-dirsize command

laktak opened the pull request:
This is a take on #152 

It adds a `calcdir` command that calculates the total size of the currently selected directories and displays it in the status and in the info column. Anything not a directory is ignored. Sorting is supported as well.

Since this is an expensive operation it is only executed when called manually.

Also the sizes for directories as reported by the OS are not very useful IMO. Can they be removed so a directory does not show a size (or 0) until it is calculated by this command?

gokcehan commented:
@laktak Thanks for working on this. This has been quite a popular request. Below are some remarks:

- I think the major issue in the current state of the patch is mixing the old directory size with this new cumulative size. I think the way forward is to remove old directory sizes from `info` until they are calculated with this new command. I'm in favor of keeping the statusline size as before as this is the same with `ls` output and shows the information from `stat` calls. Sorting should also be done without mixing. To sum up, I think `SSize` in the patch should be gone.
- Is there a reason you only calculate sizes for selections in the current directory? This is unlike many of our commands and it might be confusing for some users.
- Can we use a different name other than `calcdir`? We are not really "calculating a directory" aren't we? I think your branch name might be better for this (i.e. `dirsize`). Something resembling disk or drive usage (e.g. `diskuse` or `drive-usage`) might also be an alternative. Or maybe something with `space` in the name. Any suggestions?

I think the way this is implemented is fine for now, but below are a few further ideas for future work:

- When we calculate a directory size, technically we also calculate the directory sizes of all its subdirectories. The issue is that those subdirectories might not have been loaded so we can't simply save their sizes. So my initial idea was to use a single global map to save these directory sizes using directory paths as keys and look them up whenever required. Since I haven't actually tried this, I don't know whether or not this is a feasible approach. More specifically, it might require too much space to save all sizes. Also looking up the sizes might be computationally expensive. On the bright side, you would be able to essentially calculate everything of interest when you run the command in your home or root directories. It might also be worth adding this as a separate command from the current one.
- Somewhat in relation to above, maybe this command could also take an optional argument to calculate the size of a given path.

laktak commented:
I think I addressed your concerns in the updated PR.

- now named `get-dirsize` though I'd still say that summing everything up is calculating ;)
- dir sizes are shown as `-` (in the column) unless you run the `get-dirsize` command on them
- now uses all selected directories

I think this should cover the basics. I did not consider caching the results for subdirectories because it quickly gets very complex. We may be able to do better in a future PR (e.g. certain number of levels and files combined with a timeout that works reasonable even on slow connections ...).

pc00per commented:
I quite haven't understood how to get dirsize. `:calcdirsize` does nothing. It's still 4.0k.

laktak commented:
> I quite haven't understood how to get dirsize. `:calcdirsize` does nothing. It's still 4.0k.

You shouldn't ask for support in PRs - try https://github.com/gokcehan/lf/discussions

Also it looks like you have an old version of lf (it should show `-` initially for dirs).

pc00per commented:
Oh apologize for that
> Also it looks like you have an old version of lf (it should show - initially for dirs).

Btw I used it wrong. I had to select folder first, then it worked. Cheers...

rxmmah commented:
how to enable it globally?
